### ISSUE TITLE
403 "Forbidden" Errors while running fakestream

### ISSUE THREAD

**Issue by utmishra** [octoai/gems#11]
On running `./fakestream <config_path>`, I am receiving following errors:

```
administrators-MacBook-Pro-52:bin utkarsh.mishra$ ./fakestream /Users/utkarsh.mishra/dev/tools/octo/config/
/Users/utkarsh.mishra/dev/tools/octo/config/
Loading from Config file: /Users/utkarsh.mishra/dev/tools/octo/config//config/config.yml
Loading from Config file: /Users/utkarsh.mishra/dev/tools/octo/config//config/kong_config.yml
Loading from Config file: /Users/utkarsh.mishra/dev/tools/octo/config//config.yml
Loading from Config file: /Users/utkarsh.mishra/dev/tools/octo/config//kong_config.yml
I, [2016-08-11T16:50:52.050618 #55507]  INFO -- Octo: Octo booting up.
I, [2016-08-11T16:50:52.055021 #55507]  INFO -- Octo: I'm connected now.
I, [2016-08-11T16:50:52.060756 #55507]  INFO -- Octo: Setting callbacks.
#<Net::HTTP::Post:0x007fb33b0e42c8>
/Users/utkarsh.mishra/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/net/http/response.rb:119:in `error!': 403 "Forbidden" (Net::HTTPServerException)
    from /Users/utkarsh.mishra/.rvm/rubies/ruby-2.1.5/lib/ruby/2.1.0/net/http/response.rb:128:in `value'
    from ./fakestream:213:in `post'
    from ./fakestream:73:in `app_init'
    from ./fakestream:39:in `stream'
    from ./fakestream:59:in `stream_forever'
    from ./fakestream:233:in `_main'
    from ./fakestream:246:in `main'
    from ./fakestream:252:in `<main>'
```

During the initial setup, I guess default API credentials weren't created and fakestream must be creating a new user with credentials, but somehow its failing.

On an unrelated note, I noticed that both analytics-ws and enterprise-dashboard runs on port 9292. I will change for one, but there would be conflict for someone trying to run both at a same time.

**Comment by pranny** [comment]
Kong returns 403 because of authentication issues. You can try to run the `create_kong_config` again. Also, https://github.com/octoai/apihandler/pull/9 has been merged so a pull should help you get past some errors

**Comment by utmishra** [comment]
Still getting error:

```
Loading from Config file: /Users/utkarsh.mishra/dev/tools/octo/config//kong_config.yml
I, [2016-08-11T18:04:03.888935 #63734]  INFO -- Octo: Octo booting up.
I, [2016-08-11T18:04:03.893581 #63734]  INFO -- Octo: I'm connected now.
I, [2016-08-11T18:04:03.906671 #63734]  INFO -- Octo: Setting callbacks.
I, [2016-08-11T18:04:03.925106 #63734]  INFO -- Octo: {"data"=>[{"upstream_url"=>"http://127.0.0.1:9001/update_push_token", "strip_request_path"=>true, "request_path"=>"/update_push_token", "id"=>"a2460f8f-caf9-4e69-91ae-cbcdfe60ed28", "created_at"=>1470747335000, "preserve_host"=>false, "name"=>"UpdatePushToken"}, {"upstream_url"=>"http://127.0.0.1:9001/events", "strip_request_path"=>true, "request_path"=>"/events", "id"=>"4f54dffa-b046-49c2-8dac-cc945593ebf8", "created_at"=>1470747335000, "preserve_host"=>false, "name"=>"events"}], "total"=>2}
I, [2016-08-11T18:04:03.934539 #63734]  INFO -- Octo: Attempting to create new enterprise with name: Jaobng
W, [2016-08-11T18:04:04.039335 #63734]  WARN -- Octo: Not creating client as client name exists
I, [2016-08-11T18:04:04.039380 #63734]  INFO -- Octo: Attempting to create new enterprise with name: Octo
W, [2016-08-11T18:04:04.041195 #63734]  WARN -- Octo: Not creating client as client name exists
I, [2016-08-11T18:04:04.046561 #63734]  INFO -- Octo: Adding Plugin key-auth for api events
I, [2016-08-11T18:04:04.051039 #63734]  INFO -- Octo: Adding Plugin key-auth for api UpdatePushToken
I, [2016-08-11T18:04:04.058792 #63734]  INFO -- Octo: Adding Plugin cors for api events
I, [2016-08-11T18:04:04.061945 #63734]  INFO -- Octo: Adding Plugin cors for api UpdatePushToken
./create_kong_config.rb:207:in `block (3 levels) in create': undefined method `[]' for nil:NilClass (NoMethodError)
    from ./create_kong_config.rb:203:in `each'
    from ./create_kong_config.rb:203:in `block (2 levels) in create'
    from ./create_kong_config.rb:199:in `each'
    from ./create_kong_config.rb:199:in `block in create'
    from ./create_kong_config.rb:195:in `each'
    from ./create_kong_config.rb:195:in `create'
    from ./create_kong_config.rb:237:in `<main>'
```

**Comment by utmishra** [comment]
Hi,

Any updates on this. I am stuck with the setup because of this. 

Thanks!

**Comment by utmishra** [comment]
Also, is there any guide to create API credentials and a new User ? Fakestream would push sample data, but for production-ready setup, I would be needing such credentials. 
The current API doc has specs about pushing data, but I couldn't find any related to registering a user.

**Comment by pranny** [comment]
tl;dr
- Do a git pull for [gem-octocore](https://github.com/octoai/gem-octocore) or update `octocore` gem to version 0.0.6. You can do this by [gem update](http://guides.rubygems.org/command-reference/#gem-update). 
- Update your [config](https://github.com/octoai/config) to HEAD. You can also just update your `kong_config` as per the diff https://github.com/octoai/config/commit/91ac35d65e5dc885ae6d2fe1228c655efdcdb578 

**API Credentials and Fakestream**
- The API Credentials are auto-generated. The only place they are exposed is in the [dashboard](https://github.com/octoai/enterprise-dashboard). The dashboard also provides a JS snippet to be embedded into the web pages. For production, this is the recommended approach.
- On development environments, one can use a utility like [kong-dashboard](https://github.com/PGBI/kong-dashboard) to see the API keys and other kong details. This may prove helpful in debugging if you encounter issues related to kong.

**Adding a new user**

New user can be added by creating their corresponding entry in the `kong_config.yaml` file and running the `create_kong_config.rb` utility. This is an additive process, so it will not update anything for an existing enterprise, but will add new enterprises only.

Explanation:
Hey buddy. I was on a vacation for the long week. Sorry to keep you waiting. Turns out there was a nasty bug which we have resolved. You can find more details here. https://github.com/octoai/gem-octocore/issues/17 

Can you give it a shot now and let me know

**Comment by pranny** [comment]
Also, We are going to delete this repo soon as the gems are separated out in different issues. So, feel free to open issues there.
